### PR TITLE
Use 0 as valid value for revocation nonce

### PR DESCRIPTION
**Pull request by ilya-korotya**

### CODE REVIEW

File: src/credentials/status/agent-revocation.ts [hunk 1]
```diff
@@ -63,15 +61,23 @@ function buildRevocationMessageRequest(
...
     mtp: revocationResponse.body.mtp,
     issuer: revocationResponse.body.issuer
   };
+  if (revocationResponse.body.mtp.node_aux) {
+    rs.mtp.nodeAux = {
+      key: newHashFromBigInt(BigInt(revocationResponse.body.mtp.node_aux.key)),
+      value: newHashFromBigInt(BigInt(revocationResponse.body.mtp.node_aux.value))
+    };
+  }
+
```
Comment (vmidyllic) [inline src/credentials/status/agent-revocation.ts:81]: perhaps, you should use the code

```
const p = new Proof();
    p.existence = this.mtp.existence;
    if (this.mtp.node_aux) {
      p.nodeAux = {
        key: newHashFromBigInt(BigInt(this.mtp.node_aux.key)),
        value: newHashFromBigInt(BigInt(this.mtp.node_aux.value))
      };
    }
    const s = this.mtp.siblings.map((s) => newHashFromBigInt(BigInt(s)));

    p.siblings = [];
    p.depth = s.length;

    for (let lvl = 0; lvl < s.length; lvl++) {
      if (s[lvl].bigInt() !== BigInt(0)) {
        setBitBigEndian(p.notEmpties, lvl);
        p.siblings.push(s[lvl]);
      }
    }
    return {
      mtp: p,
      issuer: this.issuer
    };
```
you should reuse RevocationStatusDTO

File: src/credentials/status/on-chain-revocation.ts [hunk 2]
```diff
@@ -54,41 +58,59 @@ export class OnChainResolver implements CredentialStatusResolver {
...
+    const issuer = idParts[0];
+    const issuerDID = DID.parse(issuer);
+
+    const idURL = new URL(credentialStatus.id);
+
+    // if contractAddress is not present in id as param, then it should be parsed from DID
+    let contractAddress = idURL.searchParams.get('contractAddress');
+    let chainId: number;
+    if (!contractAddress) {
+      const issuerId = core.DID.idFromDID(issuerDID);
```
Comment (vmidyllic) [inline src/credentials/status/on-chain-revocation.ts:90]: use DID. Id.

File: src/storage/blockchain/chainid.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+ * Registers a new chain ID for a blockchain and network combination.
+ * If the blockchain or network is not provided, an error will be thrown.
+ * @param {string} blockchain - The blockchain name.
+ * @param {string} network - The network name.
+ * @param {number} chainId - The chain ID to be registered.
+ * @throws {Error} Throws an error if the blockchain name is not provided.
+ * @example
+ * RegisterChainId('eth', 'ropsten', 3);
+ */
+export const RegisterChainId = (blockchain: string, network: string, chainId: number): void => {
```
Comment (vmidyllic) [inline src/storage/blockchain/chainid.ts:24]: rename RegisterChainId to registerChainId

File: src/storage/blockchain/chainid.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+ * Gets the chain ID for a given blockchain and network combination.
+ * If the chain ID is not found, 0 will be returned.
+ * @param {string} blockchain - The blockchain name.
+ * @param {string} network - The network name.
+ * @returns {number} The chain ID for the specified blockchain and network.
+ * @example
+ * const chainId = GetChainId('eth', 'main');
+ * // chainId will be 1
+ */
+export const GetChainId = (blockchain: string, network: string): number => {
```
Comment (vmidyllic) [inline src/storage/blockchain/chainid.ts:45]: rename GetChainId to getChainId

File: src/storage/blockchain/chainid.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
+/**
+ * Object containing chain IDs for various blockchains and networks.
+ * @type { [key: string]: number }
+ */
+export const CHAIN_IDS: { [key: string]: number } = {
```
Comment (vmidyllic) [inline src/storage/blockchain/chainid.ts:5]: please add current chain id, so we can avoid changes to this map, as long as we can

File: src/credentials/status/on-chain-revocation.ts [hunk 6]
```diff
@@ -54,41 +58,59 @@ export class OnChainResolver implements CredentialStatusResolver {
...
+      const ethAddr = core.Id.ethAddressFromId(issuerId);
+      contractAddress = utils.getAddress(utils.hexDataSlice(ethAddr, 0));
+      const blockchain = DID.blockchainFromId(issuerId);
+      const network = DID.networkIdFromId(issuerId);
+      chainId = GetChainId(blockchain, network);
+      if (!chainId) {
+        throw new Error('chainId not found');
```
[the remark this answers is not in the document]
Comment (vmidyllic) [reply src/credentials/status/on-chain-revocation.ts:97]: chain id for $blockchain and $network is not registered

File: src/credentials/status/agent-revocation.ts [hunk 7]
```diff
@@ -63,15 +61,23 @@ function buildRevocationMessageRequest(
   };
 }
 
-function toRevocationStatus(revocationResponse: RevocationStatusResponseMessage): RevocationStatus {
+function toRevocationStatus(revocationResponse: any): RevocationStatus {
```
[the remark this answers is not in the document]
Comment (ilya-korotya) [reply src/credentials/status/agent-revocation.ts:64]: done

File: src/credentials/status/agent-revocation.ts (hunk 1 above)
Comment (ilya-korotya) [reply src/credentials/status/agent-revocation.ts:81]: done

File: src/credentials/status/on-chain-revocation.ts [hunk 8]
```diff
@@ -3,6 +3,9 @@ import { EthConnectionConfig } from '../../storage/blockchain';
 import { CredentialStatusResolver, CredentialStatusResolveOptions } from './resolver';
 import { OnChainRevocationStorage } from '../../storage/blockchain/onchain-revocation';
 import { DID } from '@iden3/js-iden3-core';
+import { core } from '../..';
```
[the remark this answers is not in the document]
Comment (ilya-korotya) [reply src/credentials/status/on-chain-revocation.ts:6]: done

File: src/credentials/status/on-chain-revocation.ts (hunk 2 above)
Comment (ilya-korotya) [reply src/credentials/status/on-chain-revocation.ts:90]: done

File: src/credentials/status/on-chain-revocation.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (ilya-korotya) [reply src/credentials/status/on-chain-revocation.ts:97]: done

File: src/storage/blockchain/chainid.ts (hunk 3 above)
Comment (ilya-korotya) [reply src/storage/blockchain/chainid.ts:24]: done

File: src/storage/blockchain/chainid.ts (hunk 4 above)
Comment (ilya-korotya) [reply src/storage/blockchain/chainid.ts:45]: done

File: src/storage/blockchain/chainid.ts (hunk 5 above)
Comment (ilya-korotya) [reply src/storage/blockchain/chainid.ts:5]: Done. We can keep state in this global variable.

File: src/credentials/status/on-chain-revocation.ts [hunk 9]
```diff
@@ -54,41 +58,59 @@ export class OnChainResolver implements CredentialStatusResolver {
...
+      contractAddress = parts[1];
     }
 
-    // TODO (illia-korotia): after merging core v2 need to parse contract address from did if `contractAddress` is not present in id as param
-    const revocationNonce = parseInt(revocationNonceParam, 10);
-    const parts = contractId.split(':');
-    if (parts.length != 2) {
-      throw new Error('invalid contract address');
+    // if revocationNonce is not present in id as param, then it should be extract from credentialStatus
+    const rv = idURL.searchParams.get('revocationNonce') || credentialStatus.revocationNonce;
```
[the remark this answers is not in the document]
Comment (ilya-korotya) [reply src/credentials/status/on-chain-revocation.ts:108]: Discussed with @Kolezhniuk . The current implementation is ok

File: src/credentials/status/agent-revocation.ts [hunk 10]
```diff
@@ -66,15 +61,23 @@ function buildRevocationMessageRequest(
...
```
[the remark this answers is not in the document]
Comment (ilya-korotya) [reply src/credentials/status/agent-revocation.ts]: outdated

**Pull request merged**